[#8165] Fix If commit and rollback fails, session close and threadLocal remove.

kitoha opened the pull request:
## Title: [#8165] [Fix] If commit and rollback fails, session close and threadLocal remove.

### What changes were proposed in this pull request?

- Sessions are properly closed even when commit/rollback operations throw exceptions

- Thread-local variables are cleared to prevent memory leaks

### Why are the changes needed?

- Memory leaks: Thread-local variables not being cleared

- Connection pool exhaustion: Unclosed connections consuming pool resources

change code 

```java
 private static void executeAndCleanupSession(Consumer<SqlSession> consumer) {
    SqlSession sqlSession = sessions.get();

    if (sqlSession == null) {
      getSessions().remove();
      return;
    }

    try (sqlSession) {
      consumer.accept(sqlSession);
    } finally {
      sessions.remove();
    }
  }
```

and modified **closeSqlSession method**

### before
```java
 public static void closeSqlSession() {
    SqlSession sqlSession = sessions.get();
    if (sqlSession != null) {
      try {
        sqlSession.close();
      } finally {
        sessions.remove();
      }
    }
}
```

### after
```java
  public static void closeSqlSession() {
    executeAndCleanupSession(session -> {});
  }
```

Fix: #8165 

### Does this PR introduce _any_ user-facing change?

### How was this patch tested?

[Commit Test](https://github.com/kitoha/gravitino/commit/5e90ddece13c43a71c323556f710dffe832492aa#diff-d1475432e21dd81c01f472b229c2049423bad8a5231cbd46afa37286c27ea54aR157)

```java 
  @Test
  public void testCommitAndCloseSqlSessionWhenCommitFails() {
    SqlSession mockSession = Mockito.mock(SqlSession.class);
    Mockito.doThrow(new RuntimeException("commit fail")).when(mockSession).commit();
    SqlSessions.getSessions().set(mockSession);
    assertThrows(RuntimeException.class, SqlSessions::commitAndCloseSqlSession);
    Mockito.verify(mockSession).close();
    assertNull(SqlSessions.getSessions().get());
  }
```

[Rollback Test](https://github.com/kitoha/gravitino/commit/5e90ddece13c43a71c323556f710dffe832492aa#diff-d1475432e21dd81c01f472b229c2049423bad8a5231cbd46afa37286c27ea54aR167)

```java
  @Test
  public void testRollbackAndCloseSqlSessionWhenRollbackFails() {
    SqlSession mockSession = Mockito.mock(SqlSession.class);
    Mockito.doThrow(new RuntimeException("rollback fail")).when(mockSession).rollback();
    SqlSessions.getSessions().set(mockSession);
    assertThrows(RuntimeException.class, SqlSessions::rollbackAndCloseSqlSession);
    Mockito.verify(mockSession).close();
    assertNull(SqlSessions.getSessions().get());
  }
```

justinmclean commented on core/src/main/java/org/apache/gravitino/storage/relational/session/SqlSessions.java:
```diff
@@ -111,4 +89,19 @@ public static void closeSqlSession() {
+      sessions.remove();
+    }
+  }
 }
```
Sorry, I missed that and just saw the close() method had been removed. You are correct, it will call close if needed.

yuqi1129 commented:
The changes look good to me. There are similar problems in `SessionUtils#doWithoutCommit` and `SessionUtils#doWithoutCommitAndFetchResult`, could you fix them all?

yuqi1129 commented:
@kitoha 

```
  public static <T, R> R doWithoutCommitAndFetchResult(Class<T> mapperClazz, Function<T, R> func) {

	// This line will open a new SqlSession if it does not exists
    T mapper = SqlSessions.getMapper(mapperClazz);

    // Use the sqlSession to perform SQL operations and DO NOT close session
    return func.apply(mapper);
  }
```

The method `doWithoutCommitAndFetchResult` and `doWithoutCommit` is designed to be wrapped with `doMultipleWithCommit` and let `doMultipleWithCommit` close the session automatically. However, in some places
like for example, `PolicyMetaService#listPoliciesForMetadataObject`, the method is not called within `doMultipleWithCommit` and will lead to connections leak. 

If you have problem with this, please let me know, then try to fix it another PR and rollback the lastes commit `[7c26c8d](https://github.com/apache/gravitino/pull/8210/commits/7c26c8dd4a3c859324e11ad53723e900952cc472)`

kitoha commented:
hi  @yuqi1129 . I'm reviewing the areas you flagged. As discussed, I'm addressing cases where doWithoutCommitAndFetchResult and doWithoutCommit are called without being wrapped by doMultipleWithCommit. I'll make the fixes and open a PR for review by tomorrow. Thanks!

kitoha commented:
hi, @yuqi1129 .  @justinmclean .

The changes have been completed. The modifications are as follows.

For doWithCommit, doWithCommitAndFetchResult, etc., on success we commit and then close the session and clear the ThreadLocal. Conversely, on failure we rollback, then close the session and clear the ThreadLocal.

doWithoutCommitAndFetchResult, doWithoutCommit, and getWithoutCommit are used in this project both as standalone methods and also under doMultipleWithCommit.
To handle both cases, I added a function to the SqlSessions class to check the current session state:

```
public static SqlSession peekSqlSession() {
  return sessions.get();
}

```

Using this function, we can check at method entry whether a session is already open. If it is open, we assume there is a prior point that uses the session, and we changed these methods not to close the session.

The method that contains this logic is shown below.

```
public static <T, R> R doWithoutCommitAndFetchResult(Class<T> mapperClazz, Function<T, R> func) {
  SqlSession existing = SqlSessions.peekSqlSession(); // load current session
  boolean owner = (existing == null);                 // determine if a session exists
  try {
    T mapper = SqlSessions.getMapper(mapperClazz);    // create if absent, otherwise reuse
    return func.apply(mapper);
  } finally {
    if (owner) {                                      // if created here, clean it up
      SqlSessions.rollbackAndCloseSqlSession();
    }
  }
}

```

doMultipleWithCommit can run operations that call getWithoutCommit-style functions, so I added logic to open a session up front. The change is below.

```
public static void doMultipleWithCommit(Runnable... operations) {
  SqlSession existing = SqlSessions.peekSqlSession();
  boolean owner = (existing == null);
  if (owner) {
    SqlSessions.getSqlSession();                      // open a session if none is open
  }

  try {
    Arrays.stream(operations).forEach(Runnable::run); // share the same session across operations
    if (owner) SqlSessions.commitAndCloseSqlSession();
  } catch (Exception e) {
    if (owner) SqlSessions.rollbackAndCloseSqlSession();
    throw e;
  }
}
```


Additionally, I added a unit test named SessionUtilsTest.
I would appreciate your review of these changes when you have time. Thank you.

kitoha commented:
hi @yuqi1129 . I’ve implemented the suggested changes. One note: getWithoutCommit (previously doWithoutCommitAndFetchResult) is used as a standalone call in many places. Updating all of these sites would result in a fairly large diff. Are you comfortable with me proceeding with that scope.

yuqi1129 commented:
> hi @yuqi1129 . I’ve implemented the suggested changes. `getWithoutCommit` (previously `doWithoutCommitAndFetchResult`) is used as a standalone call in many places. Updating all of these sites would result in a fairly large diff. Are you comfortable with me proceeding with that scope.

If so, I suggest you use a separate PR to fix it and roll back the commit to 7c26c8d. It's enough to fix the issue this PR is attached to.

yuqi1129 commented:
One thing is that please check https://github.com/apache/gravitino/pull/8242 if this PR already fixes the problem.

yuqi1129 commented:
>  When doMultipleWithCommit wraps multiple calls to doWithoutCommit (or getWithoutCommit), if getWithoutCommit closes the session and clears the ThreadLocal, the batch no longer runs under a single owner/session.

Let me double-check it. It's a big problem if it's true. We need to guarantee that `doMultipleWithCommit` is an atomic operation.

yuqi1129 commented:
> @yuqi1129 . Thanks for the quick reviews. The changes related to this PR’s issue have been merged and the problem appears resolved. However, I believe the following behavior still needs attention.
> 
> ```java
>   public static <T, R> R getWithoutCommit(Class<T> mapperClazz, Function<T, R> func) {
>     try {
>       T mapper = SqlSessions.getMapper(mapperClazz);
>       return func.apply(mapper);
>     } finally {
>       // This will decrement the counter, the session is closed only when the counter is 0.
>       SqlSessions.closeSqlSession(); // close session and remove threadlocal
>     }
>   }
> ```
> 
> When `doMultipleWithCommit` wraps multiple calls to `doWithoutCommit` (or `getWithoutCommit`), if `getWithoutCommit` closes the session and clears the `ThreadLocal`, the batch no longer runs under a single owner/session. Concretely:
> 
> ```java
> SessionUtils.doMultipleWithCommit(
>     // A ThreadLocal-bound session is created here (owner: doMultipleWithCommit)
>     () -> SessionUtils.doWithoutCommit(
>         // Uses the session created above, then closes it and clears the ThreadLocal
>         TopicMetaMapper.class, mapper -> mapper.softDeleteTopicMetasByTopicId(topicId)),
>     () -> SessionUtils.doWithoutCommit(
>         // Because the previous call cleared the ThreadLocal, a new session/ThreadLocal is created
>         OwnerMetaMapper.class,
>         mapper -> mapper.softDeleteOwnerRelByMetadataObjectIdAndType(
>             topicId, MetadataObject.Type.TOPIC.name())),
>     () -> SessionUtils.doWithoutCommit(
>         SecurableObjectMapper.class,
>         mapper -> mapper.softDeleteObjectRelsByMetadataObject(
>             topicId, MetadataObject.Type.TOPIC.name())),
>     () -> SessionUtils.doWithoutCommit(
>         TagMetadataObjectRelMapper.class,
>         mapper -> mapper.softDeleteTagMetadataObjectRelsByMetadataObject(
>             topicId, MetadataObject.Type.TOPIC.name())),
>     () -> SessionUtils.doWithoutCommit(
>         StatisticMetaMapper.class,
>         mapper -> mapper.softDeleteStatisticsByEntityId(topicId)),
>     () -> SessionUtils.doWithoutCommit(
>         PolicyMetadataObjectRelMapper.class,
>         mapper -> mapper.softDeletePolicyMetadataObjectRelsByMetadataObject(
>             topicId, MetadataObject.Type.TOPIC.name()))
> );
> ```
> 
> In this flow:
> 
> `doMultipleWithCommit` creates/binds the session in a `ThreadLocal` (becoming the lifecycle owner).
> 
> The first `doWithoutCommit`, `getWithoutCommit` uses that session but then closes it and clears the `ThreadLocal`.
> 
> Subsequent lambdas run with new sessions for `ThreadLocals`.
> 
> The final commit/close in `doMultipleWithCommit` no longer owns the original session for the earlier operations, which defeats the expectation that the whole batch is executed atomically under one session/transaction.
> 
> On the other hand, if we change `getWithoutCommit` so that it doesn’t close and clear when a session is already bound, then standalone calls to `doWithoutCommit`, `getWithoutCommit` risk leaving the `ThreadLocal` uncleared, which is also problematic.
> 
> What’s your preference here? If you think no further changes are necessary, I’m fine with closing this PR as-is. Otherwise, I’m happy to adjust the lifecycle rules. Thanks!


`doWithoutCommit` is warpped by `doWithoutCommit`, since we have use reference count mechanims to manage the sql session, only when the count is 0 then we will close the sql session. 

```
  public static void doMultipleWithCommit(Runnable... operations) {
    // This method acts as the outermost transaction boundary.
    // It increments the session count once.
    SqlSessions.getSqlSession();  // reference count = 1
    try {
      Arrays.stream(operations).forEach(Runnable::run); // refernce count then increase and decrease
      SqlSessions.commitAndCloseSqlSession(); // refernce count is 1 here, then we commit or rollback
    } catch (Exception e) {
      SqlSessions.rollbackAndCloseSqlSession();
      throw e;
    }
  }
}
```

So the whole logic seems to be okay.

kitoha commented:
@yuqi1129 . You’re right. With the reference-count mechanism, the SQL session only closes when the count reaches zero, so there’s no issue with doWithoutCommit being wrapped. I’m fine to close this PR. Thanks for your help!
